frontend/server/handlers/artifacts.ts

```diff
@@ -239,6 +253,105 @@ function getGCSArtifactHandler(options: { key: string; bucket: string }, peek: n
   };
 }
 
+function getVolumeArtifactsHandler(options: { bucket: string; key: string }, peek: number = 0) {
+  const { key, bucket } = options;
+  return async (req: Request, res: Response) => {
+    try {
+      const [pod, err] = await serverInfo.getHostPod();
+      if (err) {
+        res.status(500).send(err);
+        return;
+      }
+
+      if (!pod) {
+        res.status(500).send('Could not get server pod');
+        return;
+      }
+
+      // volumes not specified or volume named ${bucket} not specified
+      if (!Array.isArray(pod?.spec?.volumes) || !pod.spec.volumes.find(v => v?.name === bucket)) {
+        const message = `Failed to open volume://${bucket}/${key}, volume ${bucket} not configured`;
+        res.status(404).send(message);
+        return;
+      }
+
+      let serverContainer: V1Container | undefined;
+      if (Array.isArray(pod?.spec?.containers)) {
+        // container name also be called 'ml-pipeline-ui-artifact' in KFP multi user mode.
+        // https://github.com/kubeflow/manifests/blob/master/pipeline/installs/multi-user/pipelines-profile-controller/sync.py#L212
+        serverContainer = pod?.spec.containers.find(
+          c => c.name === 'ml-pipeline-ui' || c.name === 'ml-pipeline-ui-artifact',
+        );
+      }
+
+      if (!serverContainer) {
+        const message = `Failed to open volume://${bucket}/${key}, server deployment container ml-pipeline-ui or ml-pipeline-ui-artifact not found`;
+        res.status(404).send(message);
+        return;
+      }
+
+      if (!Array.isArray(serverContainer.volumeMounts)) {
+        const message = `Failed to open volume://${bucket}/${key}, volume ${bucket} not mounted`;
+        res.status(404).send(message);
+        return;
+      }
+
+      // find volumes mount
+      const volumeMount = serverContainer.volumeMounts.find(v => {
+        // volume name must be same
+        if (v?.name !== bucket) {
+          return false;
+        }
+        // if volume subPath set, volume subPath must be prefix of key
+        if (v?.subPath) {
+          return key.startsWith(v.subPath);
+        }
+        return true;
+      });
+
+      // volumes mount named ${bucket} not found
+      if (!volumeMount) {
+        const message = `Failed to open volume://${bucket}/${key}, volume ${bucket} not mounted or volume ${bucket} with subPath(which is prefix of ${key}) not mounted`;
+        res.status(404).send(message);
+        return;
+      }
+
+      // finally file path
+      const filePath = pruneAndConcatPath(volumeMount.mountPath, key, volumeMount.subPath);
+
+      const canAccess = await fs.promises
+        .access(filePath, fs.constants.F_OK | fs.constants.R_OK)
+        .then(() => true)
+        .catch(() => false);
+      if (!canAccess) {
+        res
+          .status(404)
+          .send(
+            `Failed to open volume://${bucket}/${key}, file ${filePath} not found or not readable`,
+          );
+        return;
+      }
```

Comment:
Can you explain why we need this?
It seems to me if there's something wrong with this path, we will get an error message in either `fs.promises.stat` or `fs.createReadStream` methods below.

Are those error messages not clear enough?

Reply:
yes, I think error message is clear enough, and this part can be omitted